Implement 'atomic downloads' with copy-to-temp + rename

nurpax opened the pull request:
Download a local copy into a temporary file first.  Once the download is complete, rename the temp file to the real destination.

This prevents other processes from seeing partially downloaded copies of files being downloaded.

for #3770

#### What is the purpose of this change?

Support "atomic downloads" by copy-to-temp/rename.  This prevents other processes from seeing partially downloaded files while rclone is downloading.  

#### Was the change discussed in an issue or in the forum before?

https://github.com/rclone/rclone/issues/3770

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [ ] I have added tests for all changes in this PR if appropriate.
- [ ] I have added documentation for the changes if appropriate.
- [ ] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [ ] I'm done, this Pull Request is ready for review :-)

Checklist incomplete, pushing PR for review.  If the approach looks sensible, I'll update the PR with tests & other required items.

ncw reviewed (commented):
Here are some initial thoughts. I think this might need a few iterations as operations.Copy is very complicated!

nurpax commented on fs/operations/operations.go:
```diff
@@ -374,6 +374,13 @@ func Copy(ctx context.Context, f fs.Fs, dst fs.Object, remote string, src fs.Obj
+			remoteDst := remote
+			if CanServerSideMove(f) && f.Features().PartialUploads {
+				copyRename = true
+				remoteDst = remote + ".partial-" + random.String(8)
```
I changed it to check for basename < 100 and add the partial extension only in that case.  I guess `crypt` limits are around 140 chars, so 100 might be a tad conservative.

nurpax commented on fs/operations/operations.go:
```diff
@@ -447,6 +454,22 @@ func Copy(ctx context.Context, f fs.Fs, dst fs.Object, remote string, src fs.Obj
+			// Move the copied file to its real destination path.
+			if copyRename {
+				if err == nil {
+					if doMove := f.Features().Move; doMove != nil {
```
Agreed, cleaned this up.

nurpax commented on fs/operations/operations.go:
```diff
@@ -374,6 +374,13 @@ func Copy(ctx context.Context, f fs.Fs, dst fs.Object, remote string, src fs.Obj
 		if err == fs.ErrorCantCopy {
+			copyRename := false
+			remoteDst := remote
+			if CanServerSideMove(f) && f.Features().PartialUploads {
```
Cleaned this one up in the latest commit too.

nurpax commented:
GitHub doesn't let me add an inline comment in the last comment starting with 

"I think the Move to final name should be outside of the upload retry loop, and in fact right at the end of the file, so just before here"

☝️ this is some excellent feedback, thank you!  Indeed it's a lot cleaner this way.  I implemented this in the latest comment.  I think it's pretty solid now, only one open question which I added as an inline comment into the latest revision.

If it's starting to look better, I can move on to tests & CLI options in the upcoming iterations.

nurpax commented on fs/operations/operations.go:
```diff
@@ -522,6 +512,21 @@ func Copy(ctx context.Context, f fs.Fs, dst fs.Object, remote string, src fs.Obj
+			fs.Errorf(newDst, "partial file rename failed: %v", err)
+			err = fs.CountError(err)
+			removeFailedCopy(ctx, newDst)
+			return newDst, err
```
I'm a little confused why there's `dst` and `newDst` in this function. It might be because there are a lot of assignments to `dst` and `newDst` is updated to "carry the newly created file handle forward" when there were no errors.  

I think the error handling here is correct and the debug prints print sensible output too.  I'm just not 100% sure what should be returned from this path and how the `newDst` is used by the caller.

ncw replied:
I think this looks correct.

The return from this function should be the final object in its final resting place. We try not to update `newDst` unless the function returned `err = nil` which is why I think there is `dst` and `newDst`.

I think the whole `dst` and `newDst` needs a careful review, but let's not do it in this PR! I'll do it after we merge it.

nurpax commented on fs/operations/operations.go:
```diff
@@ -447,6 +454,22 @@ func Copy(ctx context.Context, f fs.Fs, dst fs.Object, remote string, src fs.Obj
+				}
+				// TODO if the above download failed, we're left behind
+				// with a .partial-rngstring file.  What's a good way to
+				// clean it up in this retry-loop?
```
☝️ This is some excellent feedback, thank you! Indeed it's a lot cleaner this way. I implemented this in the latest comment. I think it's pretty solid now, only one open question which I added as an inline comment into the latest revision.

ncw reviewed (commented):
I think this is looking about right now :-)

Do you want to carry on with the other bits?

nurpax commented on fs/operations/operations.go:
```diff
@@ -336,6 +336,18 @@ func Copy(ctx context.Context, f fs.Fs, dst fs.Object, remote string, src fs.Obj
+			remotePartial = remote + ".partial-" + random.String(8)
+		} else {
+			// Don't make the leaf name longer if it's already lengthy to avoid
+			// trouble with file name length limits.
+			fs.Debugf(src, "Not using partial download/rename due to long file name length")
+		}
```
I've been in a kind of coffee induced brain fog.  Your suggestion is better.  I was somehow stuck on not allowing the original file name to be truncated, but that doesn't matter in this case.

nurpax replied:
I changed it a little as the above was generating a negative index.  So I'm truncating the start of the string if it's too long:

```		
remotePartial = remote + ".partial-" + random.String(8)
if len(remotePartial) > 100 {
	remotePartial = remotePartial[len(remotePartial)-100:]
}
```

ncw replied:
Remember the remote string is the whole file path (eg `dir/subdir/file.txt`)  so may have directories in. In this case truncating the start isn't a good idea!

nurpax replied:
Uhh, yeah, it absolutely isn't! I think this logic should be made to work with basenames instead of the full path.

nurpax replied:
Like this maybe:

```
	remotePartial := remote
	if !ci.Inplace && f.Features().Move != nil && f.Features().PartialUploads {
		// Avoid making the leaf name longer if it's already lengthy to avoid
		// trouble with file name length limits.
		suffix := ".partial-" + random.String(8)
		dir, base := path.Split(remote)
		baseSuffix := base + suffix
		if len(baseSuffix) > 100 {
			remotePartial = path.Join(dir, baseSuffix[len(baseSuffix)-100:])
		} else {
			remotePartial = remote + suffix
		}
	}
```

nurpax commented:
@ncw Should the `--inplace` flag be a global flag and ignored on backends where this is not needed/unsupported, or should each supporting backend have their on flag like `--local-inplace`?  I'd vote for the former: single global `--inplace` option.

ncw commented:
> @ncw Should the `--inplace` flag be a global flag and ignored on backends where this is not needed/unsupported, or should each supporting backend have their on flag like `--local-inplace`? I'd vote for the former: single global `--inplace` option.

Yes this needs to be a global flag as it affects `operations.Copy` which is used by everything.

BTW You can disable a feature for a backend, eg `--disable PartialWrites` which would be equivalent to `--inplace=true`.

nurpax commented:
OK, I added --inplace and a couple of tests.  I did notice that some tests were failing at least `TestSyncUTFNorm`, I didn't check yet if those were earlier failures or caused by my change.

nurpax commented:
I find that TestSyncUTFNorm is fixed with this change:

```
diff --git a/fs/operations/operations.go b/fs/operations/operations.go
index 11c4b4795..fbe39cbd6 100644
--- a/fs/operations/operations.go
+++ b/fs/operations/operations.go
@@ -337,7 +337,7 @@ func Copy(ctx context.Context, f fs.Fs, dst fs.Object, remote string, src fs.Obj
        hashType, hashOption := CommonHash(ctx, f, src.Fs())

        remotePartial := remote
-       if !ci.Inplace && f.Features().Move != nil && f.Features().PartialUploads {
+       if !ci.Inplace && f.Features().Move != nil && f.Features().PartialUploads && dst == nil {
                // Avoid making the leaf name longer if it's already lengthy to avoid
                // trouble with file name length limits.
                suffix := ".partial-" + random.String(8)
```

Ie., don't do download/rename if a non-nil dst fileobject is passed to `Copy()`.  In a normal "rclone copy" operation `dst` will be `nil` upon entry to this function, but I don't know in which situations a valid `dst` is passed in.

What seems to happen in this test is that the `Move()` at the end moves a `newDst` object to the remote name.  But somehow `newDst` in this case points to the original file, not the suffixed temporary file.  There's probably a case along the way where newDst is not correctly updated to point to the suffix name.

ncw commented:
I think the problem is this...

If `dst != nil` then you want to be using `dst.Remote()` to name the file, not `src.Remote()`.

Thats a subtle point also... Some backends do different things on Update vs Put. If you use Update when you should have used Put on Google Drive you'll make a duplicate file.

Now on the backends we are talking about, `local`, `sftp`, `ftp` etc they don't do anything special. Though perhaps we'll be missing out on special permissions on the file which would have been preserved if we updated the existing file rather than made a new one.

ncw commented:
> I can get tests (ie., `cd fs/sync && go test`) to pass with this change: [4d31ccf](https://github.com/rclone/rclone/commit/4d31ccf60071053c1e879dfc30ac58f38f77eb65)
> 
> But to be fair, I don't completely understand this function so it's hard to tell if that's the right thing to do. :)

I think that patch is nearly right - I commented on it what I think it should be.

nurpax commented:
Added a comment here: https://github.com/rclone/rclone/commit/4d31ccf60071053c1e879dfc30ac58f38f77eb65#commitcomment-109940910 (I'm finding it really hard to keep track of review comments in GitHub PRs :))

jakanaka commented:
How this is supposed to work? I mean what flags to use?

nurpax commented:
The patch implements this so that it's enabled by default.  If you want to revert to old behavior, you can turn it off with the "--inplace" flag.  This is the same default and same flag as in rsync.

jakanaka commented:
> The patch implements this so that it's enabled by default. If you want to revert to old behavior, you can turn it off with the "--inplace" flag. This is the same default and same flag as in rsync.

What does it do exactly? what will happen if i do this?


`rclone copy "/mnt/folder1" "sftp:/folder1"`

here sftp is a sftp remote

ncw commented:
@nurpax I think next week I will pull this locally and see if I can find out what is going on. If I manage to fix it I'll push a commit on top of yours.

ncw commented:
@nurpax I've spent quite a long time working on this. I've

- rebased the commits and fixed the conflicts
- squashed the commits into two main commits
- fixed the problems (these were to do with calling the Update method)
- added some more commits with more fixes

So I think this is nearly done!

It needs

- documentation (I'm happy to write this)
- decision on whether to set the feature flag for FTP/SFTP

Can you pull what I've done (I'd suggest into a new branch since I rebased it) and see how it is working for you?

nurpax commented:
@ncw, thank you so much for helping out with this change!  I read through your modifications and it locally and it works as I'd expect it to work: copying new and replacing existing files works, I tried with a bunch of different "rclone copy/sync" scenarios where the source was modified.  Tests also passed.

As far as my original needs for this feature go, it's done. 👍   I don't have an opinion on ftp/sftp.  We don't use that path and I have little recent experience on using ftp, so I can't tell what downsides enabling this might have.

----

For the record, there are some requests in https://github.com/rclone/rclone/issues/3770#issuecomment-789356722 that are not implemented in this change.  I think they could hide the partial files by masking out `*.partial` in their web server config.  The `.partial` would have to become part of the "official rclone interface" then.  A simple workaround (at the cost of extra diskspace) would be to rclone into a local directory (outside of the web server dirs) first, then perform another rclone sync from the local directory to the web server dir.

The comment mentions `--partial-dir` as a way to specify a separate directory for the temp downloads but it seems rsync's `--partial-dir` doesn't quite work like that.  See e.g., this thread: [like copying `--partial-dir` to rclone ](https://www.reddit.com/r/linuxquestions/comments/x6gl0s/rsync_ignoring_partialdir/)

It's possible to revert back to old behavior by using the `--inplace` flag, so IMO merging what's done now is definitely forward progress anyway. :)

ncw commented:
@nurpax 
> thank you so much for helping out with this change! I read through your modifications and it locally and it works as I'd expect it to work: copying new and replacing existing files works, I tried with a bunch of different "rclone copy/sync" scenarios where the source was modified. Tests also passed.

Great!

> As far as my original needs for this feature go, it's done. +1 I don't have an opinion on ftp/sftp. We don't use that path and I have little recent experience on using ftp, so I can't tell what downsides enabling this might have.

I think enabling this on SFTP is pretty safe. Enabling this on FTP is probably less safe as there are a lot of extremely crusty FTP servers out there.

I'm going to enable it for FTP and SFTP since the integration tests all pass. I had to patch the SFTP backend to allow overwriting files when renaming.

> For the record, there are some requests in [#3770 (comment)](https://github.com/rclone/rclone/issues/3770#issuecomment-789356722) that are not implemented in this change. I think they could hide the partial files by masking out `*.partial` in their web server config. The `.partial` would have to become part of the "official rclone interface" then. A simple workaround (at the cost of extra diskspace) would be to rclone into a local directory (outside of the web server dirs) first, then perform another rclone sync from the local directory to the web server dir.

I'm happy for `*.partial` to become part of the official rclone interface - I've written it into the docs.

> The comment mentions `--partial-dir` as a way to specify a separate directory for the temp downloads but it seems rsync's `--partial-dir` doesn't quite work like that. See e.g., this thread: [like copying `--partial-dir` to rclone ](https://www.reddit.com/r/linuxquestions/comments/x6gl0s/rsync_ignoring_partialdir/)

Yes `--partial-dir` on rsync does something a little different. However I think it is probably close enough if we wanted to use it for the purpose of storing `*.partial` files.

> It's possible to revert back to old behavior by using the `--inplace` flag, so IMO merging what's done now is definitely forward progress anyway. :)

Yes!

I've just pushed an update with docs, sftp and ftp support. If that builds OK then I'll merge :-)

Thank you for doing this, and in particular thank you for untangling the confusion re rsync as to exactly what this feature should do.

nurpax commented:
Excellent!  Thanks for the help, I'm glad it's merged! :)

Jwink3101 commented:
This can potentially close #6206 and #6284

kapitainsky commented:
Hi guys. Please have a look at:

https://forum.rclone.org/t/local-to-local-with-links-drops-the-rclonelink-extension/39483

Looks like this PR introduced some small bug.

ncw commented:
I will look at this shortly!
